src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
+                // var r = new R() { field = ref var() };
+                Diagnostic(ErrorCode.ERR_VarInvocationLvalueReserved, "var()").WithLocation(2, 31),
+                // (2,31): error CS0103: The name 'var' does not exist in the current context
+                // var r = new R() { field = ref var() };
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "var").WithArguments("var").WithLocation(2, 31)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_ReadonlyRefField()
+        {
+            var source = @"
+int x = 42;
+var r = new R() { field = ref x };
+
+ref struct R
+{
+    public readonly ref int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,19): error CS0191: A readonly field cannot be assigned to (except in a constructor or init-only setter of the type in which the field is defined or a variable initializer)
+                // var r = new R() { field = ref x };
+                Diagnostic(ErrorCode.ERR_AssgReadonly, "field").WithLocation(3, 19)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_RefReadonlyField()
+        {
+            var source = @"
+int x = 42;
+var r = new R() { field = ref x };
+
+ref struct R
+{
+    public ref readonly int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RefInitializer_RefReadonlyValue_Field()
+        {
+            // If the left operand is a writeable ref (i.e. it designates anything other than a `ref readonly` local or  `in` parameter), then the right operand must be a writeable lvalue.
+            var source = @"
+class C
+{
+    ref readonly int Value() => throw null;
+
+    void M()
+    {
+        var r = new R() { field = ref Value() };
+    }
+}
+
+ref struct R
+{
+    public ref int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,39): error CS8331: Cannot assign to method 'C.Value()' because it is a readonly variable
+                //         var r = new R() { field = ref Value() };
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "Value()").WithArguments("method", "C.Value()").WithLocation(8, 39)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_AssignInParameter()
+        {
+            var source = @"
+class C
+{
+    static void F(in int i)
+    {
+        _ = new R1 { _f = ref i };
+        _ = new R2 { _f = ref i }; // 1
+    }
+}
+
+ref struct R1
+{
+    public ref readonly int _f;
+}
+
+ref struct R2
+{
+    public ref int _f;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (7,31): error CS8331: Cannot assign to variable 'in int' because it is a readonly variable
```

Comment:
ideally, the diagnostic message would say `in int i` here.

Reply:
https://github.com/dotnet/roslyn/issues/62096